pep-0635.rst

```diff
@@ -0,0 +1,689 @@
+to 'function bodies' or actions.  This general mapping structure can be
+found in different context as well.  Algol-derived languages usually
+provide a switch table that maps ordinal values to actions, whereas Lisp
+has a more general mapping from general conditions to actions.  Although
+all these constructs share a similar overall structure and some syntax,
+their intents and motivation differs highly.  In particular, pattern
+matching as proposed here is not intended as or an extension of a switch
+structure, although it is possible to emulate it to a large degree with
+the syntax proposed here.
+
+
+
+Rationale
+=========
+
+TBD.
+
+This section should provide the rationale for individual design decisions.
+It takes the place of "Rejected ideas" in the standard PEP format.
+It is organized in sections corresponding to the specification (PEP 634).
+
+
+Overview and terminology
+------------------------
+
+
+
+The ``match`` statement
+-----------------------
+
+TBD.
+
+The overall syntax of each case clause is similar to that of lambda
+functions, although the body of case clauses are blocks of statements
+rather than expressions.  Compare, for instance a lambda function to
+add two values::
+
+  lambda x, y: x + y
+  
+with a case clause performing the same operation::
+
+  case x, y:
+      return x + y
+      
+The case clause would, of course, be embedded in a match statement and
+ultimately in a function.  Nonetheless, understanding the patterns
+following the ``case`` as a generalisation of parameters is a solid
+mental model to approach and understand pattern matching.
+
+
+Match semantics
+~~~~~~~~~~~~~~~
+
+
+
+.. _patterns:
+
+Patterns
+--------
+
+Patterns are most aptly described as a generalisation of parameters as
+in function definitions.  They also share some characteristics with
+targets of iterable unpacking.  Most importantly, however, patterns are
+not expressions.  A pattern cannot be evaluated or executed, it is a
+static declaration of a structural template.  This declarative nature
+is a characteristic it shares with ``global`` statements, for instance,
+but also with regular expressions or context-free grammars.
+
+Python's iterable unpacking can assign values to any valid target,
+including attributes and subscripts.  This allows you to write, e.g.,
+``self.x, self.y = x, y`` in a class' initialisator, or
+``a[i+1], a[i] = a[i], a[i+1]`` to swap two elements in a list.  The
+same approach, however, does not work for patterns due to their
+'conditional' nature.  It is at the very core of pattern matching that
+a pattern may safely fail to match a given subject and reject it.  In
+order for this to make sense and to reason about patterns, it is
+imperative to avoid any side effects (as far as possible within the
+bounds of a dynamic language).  Patterns can therefore not assign
+values to arbitrary targets, but rather bind *local* variables to
+values extracted from the data provided.
```

Comment:
I fully understand your reservations.  It is certainly true that we cannot have something that is entirely static and declarative in a dynamic language.

On the other hand, if I understood correctly, you and Brandt just argued in [issue 154](https://github.com/gvanrossum/patma/issues/154) that you should expect that your code breaks if you do funny stuff with side effects and all that.  So, at least in some cases we seem to take a pragmatic stance in that all those examples (name lookup, `__eq__`, etc.) _are without side effects_, unless you specifically invent a fancy implementation that adds them.

Don't we have a similar issue when sorting a list, say?  Because the sorting algorithm/implementation may change, there is no guarantee how often the compare methods on each element will be called, anyway.  Hence, if you come up with an implementation where comparing two objects has strange side effects, you cannot expect sort to guarantee that it works correctly...

So, even though there are holes in the abstraction, I feel like this is still an important guiding principle.  And it tells you that if you do funny stuff that violates our assumption, any breakage of the system is on you.

Reply:
The key difference is that in order to generate efficient code for `match` in a static language, the compiler can look at the class hierarchy, and if the classes are sealed, you can generate a switch table.

But the Python compiler cannot even trust the class hierarchy in the current module (there are lots of hacks that play games with this, e.g. metaclasses, decorators and monkey-patching), and it truly cannot know anything about the class hierarchy in other modules. Even builtins are dicey.

We can generate bytecode that caches the outcome of e.g. `a.b.C` if it occurs in multiple patterns, but the cache would have to be rebuilt each time we enter the same `match` -- even our wildest undefined behavior does not allow us to assume that `a.b.C` will still be the same thing the next time. Ditto for the other things we might cache, like `getattr()` or `__eq__`.

Reply:
> [...] if the classes are sealed, you can generate a switch table.

I am not entirely sure this is what really happens.  At least, I have not come across a compiler that actually does this.  Rather, the things that I have seen were all based on the decision tree (which is why I brought it up), which would also work for us.

This is probably one of the differences to the visitor pattern: branching on the class alone is not that common in pattern matching.  Except for a lot of cases with just two or three classes involved in the first place (e.g. Scala's "optionals" `Some(x)` vs. `None`, linked lists with `List(head, tail)` vs. `Nil`, trees with `Node(...)` vs. `Leaf`, etc.).  But in these cases, you don't need a switch table, either, as a simple `if` does the job at least as good.

Scala, for instance, compiles a pattern matching expression down to a switch table if and only if all patterns are literal patterns.  I recently had a case where I used named constants instead of putting in all the numeric values directly, and was quite disappointed to find that they compile that to a huge if-elif-else-chain!

I have a feeling that switch tables really start to become interesting and make a difference performance-wise when you have so many possible cases involved that you would not design them as sealed classes anymore.